sensors: Add streaming APIs

yperess opened the pull request:
Introduce a streaming API that uses the same data path as the async API.

This includes features to the decoder:
* Checking if triggers are present

Adding streaming features built ontop of existing triggers:
* Adding 3 operations to be done on a trigger
  * include - include the data with the trigger information
  * nop - do nothing
  * drop - drop the data (flush)
* Add a new sensor_stream() API to mirror sensor_read() but add an
optional handler to be able to cancel the stream.

Fixes #1387 
Fixes #13718

Commit:
sensors: Add streaming APIs

Introduce a streaming API that uses the same data path as the async API.

This includes features to the decoder:
* Checking if triggers are present

Adding streaming features built ontop of existing triggers:
* Adding 3 operations to be done on a trigger
  * include - include the data with the trigger information
  * nop - do nothing
  * drop - drop the data (flush)
* Add a new sensor_stream() API to mirror sensor_read() but add an
optional handler to be able to cancel the stream.

Signed-off-by: Yuval Peress <peress@google.com>
topic#sensor_stream

Commit:
icm42688: Implement streaming APIs

Add streaming implementation for icm42688 using both threshold and
full FIFO triggers.

Signed-off-by: Yuval Peress <peress@google.com>
topic#sensor_stream

asemjonovs commented on include/zephyr/drivers/sensor.h:
```diff
@@ -246,6 +246,9 @@ enum sensor_trigger_type {
 	/** Trigger fires when no motion has been detected for a while. */
 	SENSOR_TRIG_STATIONARY,
+
+	SENSOR_TRIG_FIFO_THRESHOLD,
+	SENSOR_TRIG_FIFO_FULL,
```
Nit: Add comment

yperess replied:
Done

teburd commented:
I'll try and look at this soon

nashif commented:
@yperess  do we need to update the documnetaion with this new addition?

yperess commented:
> @yperess  do we need to update the documnetaion with this new addition?

I'm holding off on the documentation to make sure it's stable before we start directing people to it. Though if you think there's a benefit to it I'm happy to do it and put a banner above for "experimental"

I have some changes in the pipe for the decoder (expected to be done late next week)

yperess commented:
@teburd and @asemjonovs when looking at the `Compare` the first force push was a rebase, the second adds the doxygen.

yperess commented:
Had an issue when rebasing. I had to move the streaming logic out to a separate file.

MaureenHelm commented on drivers/sensor/sensor_shell.c:
```diff
@@ -349,21 +443,32 @@ static int cmd_get_sensor(const struct shell *sh, size_t argc, char *argv[])
 	return 0;
 }
 
+static void sensor_shell_processing_entry_point(void *a, void *b, void *c)
+{
+	ARG_UNUSED(a);
+	ARG_UNUSED(b);
+	ARG_UNUSED(c);
+
+	while (true) {
+		sensor_processing_with_callback(&sensor_read_rtio,
+						sensor_shell_processing_callback);
+	}
+}
+K_THREAD_DEFINE(sensor_shell_processing_tid, CONFIG_SENSOR_SHELL_THREAD_STACK_SIZE,
+		sensor_shell_processing_entry_point, NULL, NULL, NULL, 0, 0, 0);
```
I know this is useful for testing the new streaming feature, but there are other users of the sensor shell that don't want to take the memory hit. Can you please partition the sensor shell changes so the it can still be used with `CONFIG_SENSOR_ASYNC_API=n`?

yperess replied:
Done

MaureenHelm commented on include/zephyr/drivers/sensor.h:
```diff
@@ -246,6 +246,12 @@ enum sensor_trigger_type {
 	SENSOR_TRIG_STATIONARY,
+
+	/** Trigger fires when the FIFO threshold has been reached. */
+	SENSOR_TRIG_FIFO_THRESHOLD,
```
Replace "threshold" with "watermark" to be consistent with the attribute enum?

yperess replied:
Done

yperess commented:
Build error is caused by https://github.com/zephyrproject-rtos/zephyr/pull/60695

MaureenHelm commented on drivers/sensor/sensor_shell.c:
```diff
@@ -233,19 +263,43 @@ static int parse_sensor_value(const char *val_str, struct sensor_value *out)
-					     void *userdata)
+void sensor_shell_processing_callback(int result, uint8_t *buf, uint32_t buf_len, void *userdata)
 {
+	static struct {
+		int64_t accumulator;
+		int8_t shift;
+		int count;
+		bool is_active;
+	} sample_stats[SENSOR_CHAN_COMMON_COUNT];
```
This might be better done with the stats subsys

lixuzha commented:
Hi @yperess,

I'm evaluating how to enable sensing subsys with stream api based on this pull request.
During the evaluation, I thought there might be some potential gaps in this PR. Could you please help check and comment?

1. Can ACCEL FIFO and GYRO FIFO be enabled separately? For example, when the user only wants to use the Accel function, turning off Gyro can save more power.

2. If an A/G combo sensor is upgraded with streaming API, can user enable accel with FIFO, while keep using gyro in polling mode?

yperess commented:
> Hi @yperess,
> 
> I'm evaluating how to enable sensing subsys with stream api based on this pull request.
> During the evaluation, I thought there might be some potential gaps in this PR. Could you please help check and comment?
> 
> 1. Can ACCEL FIFO and GYRO FIFO be enabled separately? For example, when the user only wants to use the Accel function, turning off Gyro can save more power.

There's only 1 hardware FIFO usually. The inclusion depends on the sample rate. So by enabling the FIFO + setting an accelerometer ODR you're including the accelerometer in the FIFO. If you then set the ODR for the gyro, it'll be included too.

> 2. If an A/G combo sensor is upgraded with streaming API, can user enable accel with FIFO, while keep using gyro in polling mode?

You can always poll and get the latest sample. So the polling result should always be the same as the tail of the FIFO.

teburd commented on include/zephyr/drivers/sensor.h:
```diff
@@ -511,13 +528,38 @@ struct sensor_decoder_api {
 	const struct device *sensor;
-	enum sensor_channel *const channels;
+	const bool is_streaming;
+	union {
```
I don't think this should be a union, the reader may wish to select channels and trigger sources independently, e.g. on data ready read the gyro x and y channels only

yperess replied:
I wanted to avoid adding things that would/could get confusing. For example, lets say my application wants to know about the `RESET_DONE_INT`. I would have to add it to the triggers in sensor.h, then I could set the stream to tell me about it. I wouldn't set any channels. Does that mean I want to turn off the FIFO?

The current design is such that the read configs are minimally intrusive. Setting filters like that which affect the configuration registers should be done via attributes. In your example specifically:
* If the application only cares about reading the gyro data they can set the accel ODR to 0Hz
* If the application wants to batch gyro and poll accel they can set the ODR for both and add a PR to add new attributes to configure the FIFO to exclude channels.

teburd reviewed (commented):
Some notes on places that maybe are worth changing to account for something pointed out by @lixuzha (great eye!)

yperess commented on include/zephyr/drivers/sensor.h:
```diff
@@ -540,12 +582,24 @@ struct sensor_read_config {
 	};                                                                                         \
 	RTIO_IODEV_DEFINE(name, &__sensor_iodev_api, &__sensor_read_config_##name)
 
+#define SENSOR_DT_STREAM_IODEV(name, dt_node, ...)                                                 \
```
Assumption: I think this isn't a super common thing (i.e. many sensors won't have this filter).

Overall, that feature sounds like we should be adding an attribute. This way the application can configure the FIFO to filter gyroscope data, and the steam request can focus on the when aspect. can you link a datasheet so I can see this filter?

teburd replied:
icm42688, icm20469, lsm6dso, bmi270 all support this sort of thing where you can select what goes in the fifo

yperess replied:
OK, found it in `FIFO_CONFIG1` for the icm42688, thanks. I still think this belongs more as an attribute than as a part of the read request no?

teburd replied:
I guess that's an interesting question here, what constitutes the stream options (sensor read config) vs sensor attributes. At the moment the poll interface lets you select channels and in this PR the stream interface lets you select which triggers but not which channels.

That seems a little strange I guess. I'm not opposed to making them attributes, but I guess if the goal is uniform interfaces for simplicity, it would make sense to allow the stream to also select which channels to read.

lixuzha replied:
I kind of agree with Tom.

yperess replied:
The problem I'm having with specifying the channels in this request is ambiguity. Now, please note I'm very possibly wrong here, but the point I'm making is that it's possible to make the following assumption because that's what I would think.

If I were implementing a driver and I got a steam request for the watermark trigger and just accel (no gyro). I would assume that means I can turn off the gyro. Now, the discussion we're having says no, just remove it from the FIFO. But that's not really specified in the API and it's open for interpretation.

Alternatively, if we add a new attribute `SENSOR_ATTR_FIFO_ACCEL_ENABLE` and the same for gyro, then there's no confusion. The steam option just says enable these triggers and give me any associated data. You'll always get the triggers with the timestamp, and it gives you 3 options of what to do with the rest of the data:
* Discard / flush
* Ignore / do nothing
* Read

There's a very clear line between what the API is trying to do. Anything more, you need to call sensor_set_attr

lixuzha commented:
> > Hi @yperess,
> > I'm evaluating how to enable sensing subsys with stream api based on this pull request.
> > During the evaluation, I thought there might be some potential gaps in this PR. Could you please help check and comment?
> > 
> > 1. Can ACCEL FIFO and GYRO FIFO be enabled separately? For example, when the user only wants to use the Accel function, turning off Gyro can save more power.
> 
> There's only 1 hardware FIFO usually. The inclusion depends on the sample rate. So by enabling the FIFO + setting an accelerometer ODR you're including the accelerometer in the FIFO. If you then set the ODR for the gyro, it'll be included too.
> 
From the code, looks like every time when sensor ODR is changed, the FIFO will need to be clean up and reset. 
Based on upper layer virtual sensor and application request, the sensing subsystem arbitration logic could change gyro ODR back and forth frequently, but in proposed design, gyro ODR change will have an impact to accel FIFO status, which does not look efficient and reasonable to me.

> > 2. If an A/G combo sensor is upgraded with streaming API, can user enable accel with FIFO, while keep using gyro in polling mode?
> 
> You can always poll and get the latest sample. So the polling result should always be the same as the tail of the FIFO.

The question is actually not about how to poll gyro data. My question is that: if I want to config accel in FIFO mode (suppose user wants to stream accel in extremely high ODR, and maximally leverage the FIFO), and config gyro in polling mode (with normal ODR), can I do it with current design?


Furthermore, suppose a user developed an application based on the new stream API for channel accel_x, y, z and gyro_x, y, z, with two separated sensors (bma280 + bmg160), then the user change his sensor BOM to a combo sensor (icm42688). In this case, does this user need to upgrade his application?

yperess commented:
> From the code, looks like every time when sensor ODR is changed, the FIFO will need to be clean up and reset. Based on upper layer virtual sensor and application request, the sensing subsystem arbitration logic could change gyro ODR back and forth frequently, but in proposed design, gyro ODR change will have an impact to accel FIFO status, which does not look efficient and reasonable to me.

A few things on that,
1. I don't imagine the arbitrated configuration changing enough to make a difference
2. @teburd and I are working on a vertical decoder which will replace the one we have. This will mean that the sensor.h data representation will match 1:1 with that of the subsystem. i.e. the user can ask to decode all the frames of channel X. This allows us to move the structs like `sensing_sensor_three_axis_data` down one level to `sensor_three_axis_data` which will be the output of the decoder. This will include the timestamp and offset and will get rid of exactly what you're describing (having to flush the FIFO on configuration changes).

Hopefully the second point here helps.

> The question is actually not about how to poll gyro data. My question is that: if I want to config accel in FIFO mode (suppose user wants to stream accel in extremely high ODR, and maximally leverage the FIFO), and config gyro in polling mode (with normal ODR), can I do it with current design?

I'd like to consolidate these threads as it seems we talk about the same thing in 3 different places. Please continue this question on https://github.com/zephyrproject-rtos/zephyr/pull/60063#discussion_r1285059486. The short of it, send a PR to add this FIFO config as an attribute.
 
> Furthermore, suppose a user developed an application based on the new stream API for channel accel_x, y, z and gyro_x, y, z, with two separated sensors (bma280 + bmg160), then the user change his sensor BOM to a combo sensor (icm42688). In this case, does this user need to upgrade his application?

Yes, but that's exactly why we're building the subsystem. Every layer has it's own job. We can't expect the kernel level drivers to abstract these board layout choices. My advice to the developer would be to use the sensing subsystem. One of the advantages of pushing for the vertical decoder mentioned above is that the data representation will be the same for sensor.h and sensing.h. This means that once in and people are migrating their existing applications to use sensor.h's new API, they'll also have an easy time migrating to sensing.h.

lixuzha commented:
> > Furthermore, suppose a user developed an application based on the new stream API for channel accel_x, y, z and gyro_x, y, z, with two separated sensors (bma280 + bmg160), then the user change his sensor BOM to a combo sensor (icm42688). In this case, does this user need to upgrade his application?
> 
> Yes, but that's exactly why we're building the subsystem. Every layer has it's own job. We can't expect the kernel level drivers to abstract these board layout choices. My advice to the developer would be to use the sensing subsystem. One of the advantages of pushing for the vertical decoder mentioned above is that the data representation will be the same for sensor.h and sensing.h. This means that once in and people are migrating their existing applications to use sensor.h's new API, they'll also have an easy time migrating to sensing.h.

Please be aware that sensing subsystem is optional module in Zephyr, and in above case, we are talking about the application directly uses sensor device driver without sensing subsystem. My understanding is that this should be the default case whenever we consider upgrading sensor device driver API, right?

yperess commented:
> > Yes, but that's exactly why we're building the subsystem. Every layer has it's own job. We can't expect the kernel level drivers to abstract these board layout choices. My advice to the developer would be to use the sensing subsystem. One of the advantages of pushing for the vertical decoder mentioned above is that the data representation will be the same for sensor.h and sensing.h. This means that once in and people are migrating their existing applications to use sensor.h's new API, they'll also have an easy time migrating to sensing.h.
> 
> Please be aware that sensing subsystem is optional module in Zephyr, and in above case, we are talking about the application directly uses sensor device driver without sensing subsystem. My understanding is that this should be the default case whenever we consider upgrading sensor device driver API, right?

Right, that's inline with what I said. If the data representation of the kernel level drivers is the same though, then it's easy for developers to start off managing the sensors themselves and migrate to adding the subsystem when they want:
* Arbitration
* Multiplexing
* Virtual sensors

The idea is that the application logic that's handling the sensor data should require minimal changes when adding the subsystem. But this seems to be a discussion more for https://github.com/zephyrproject-rtos/zephyr/issues/36963

KeHIntel commented:
> it's easy for developers to start off managing the sensors themselves and migrate to adding the subsystem when they want: Arbitration, Multiplexing, Virtual sensors

Hi, Yuval, you are definitely right that Zephyr application based on sensing subsystem with wrapper over streaming API should have no such problem, but **Lixu's question here is more focusing on streaming API itself**. 

Streaming API can be directly used by Zephyr application without sensing subsystem, we will need to consider impact to this kind of cases as well. I think current expectation is that when sensor channels have no change, sensor application should have no change, even sensor BOM is changed.

nashif commented:
> Hi, Yuval, you are definitely right that Zephyr application based on sensing subsystem with wrapper over streaming API should have no such problem, but **Lixu's question here is more focusing on streaming API itself**.
> 
> Streaming API can be directly used by Zephyr application without sensing subsystem, we will need to consider impact to this kind of cases as well. I think current expectation is that when sensor channels have no change, sensor application should have no change, even sensor BOM is changed.

@yperess  can you please address the question above?

yperess commented:
> > it's easy for developers to start off managing the sensors themselves and migrate to adding the subsystem when they want: Arbitration, Multiplexing, Virtual sensors
> 
> Hi, Yuval, you are definitely right that Zephyr application based on sensing subsystem with wrapper over streaming API should have no such problem, but **Lixu's question here is more focusing on streaming API itself**.
> 
> Streaming API can be directly used by Zephyr application without sensing subsystem, we will need to consider impact to this kind of cases as well. I think current expectation is that when sensor channels have no change, sensor application should have no change, even sensor BOM is changed.

I believe what I'm understanding is that the concern is that we shouldn't be taking into account the subsystem when designing the kernel drivers.

If that's correct, I'd say sure... maybe... but does it really matter? If the data representation is optimal for data processing, then it's the best representation. Regardless of where it is (subsystem or driver). If it's the best representation, then it belongs in both. The current representation of channels is overly restrictive. It assumes a single dimension per channel (with a few odd `XYZ` exceptions). I don't think this is necessary and I believe we can do better. In the new APIs (the async APIs) the channel is much more like a type. The type maps to a struct. Custom sensor types/channels in the application may define new structs as long as they provide a decoder along with it.

What does this look like? It might mean that a sensor can produce much more rich data. Things like a heatmap or a lidar scan even.

KeHIntel commented:
> I believe what I'm understanding is that the concern is that we shouldn't be taking into account the subsystem when designing the kernel drivers.

I think so, this is today's default model to use sensor in Zephyr, and sensing subsystem is not fully merged. Any change to device driver API shall consider the default usage model at first place.
 
> If that's correct, I'd say sure... maybe... but does it really matter? 

It matters. My understanding is that Zephyr sensors has a market in low end IOT devices whose sensing requirements are simple, but memory/computation requirements are strict. I foresee that these devices will continue adopt the default model to use sensors, without sensing subsystem. They may adopt FIFO, as it brings a few benefits, so we have to consider them in this PR.

yperess commented:
> > I believe what I'm understanding is that the concern is that we shouldn't be taking into account the subsystem when designing the kernel drivers.
> 
> I think so, this is today's default model to use sensor in Zephyr, and sensing subsystem is not fully merged. Any change to device driver API shall consider the default usage model at first place.
> 
> > If that's correct, I'd say sure... maybe... but does it really matter?
> 
> It matters. My understanding is that Zephyr sensors has a market in low end IOT devices whose sensing requirements are simple, but memory/computation requirements are strict. I foresee that these devices will continue adopt the default model to use sensors, without sensing subsystem. They may adopt FIFO, as it brings a few benefits, so we have to consider them in this PR.

Right but we're just talking about data representation here. I'm not saying they need to use the subsystem.

KeHIntel commented:
> Right but we're just talking about data representation here. I'm not saying they need to use the subsystem.

I agree with your points on the data representation, your effort in driving FIFO streaming and vertical decoder are well appreciated. However, our comments on this PR is mainly about the modification of the sensor device driver API itself. With driver API today, an application can work with A/G combo sensor, as well as A/G separated sensor; however, with this PR, this will be broken, and I believe this shall be fixed before we merge this PR.

Commit:
sensors: convert fifo_wm to batch_dur

Having a % FIFO watermark isn't very useful as it doesn't convey how long
the SoC can sleep (or do other work) while batching sensor data. Convert
the attribute to a batch duration using ticks. Currently the ticks are
in system ticks, but eventually when an external clock is attached to
the sensor it will be in the external clock's ticks.

Signed-off-by: Yuval Peress <peress@google.com>

lixuzha commented on include/zephyr/drivers/sensor.h:
```diff
@@ -567,12 +611,24 @@ struct sensor_read_config {
+	static struct sensor_read_config __sensor_read_config_##name = {                           \
+		.sensor = DEVICE_DT_GET(dt_node),                                                  \
+		.is_streaming = true,                                                              \
+		.triggers = __trigger_array_##name,                                                \
```
The type of `.triggers` is different form `__trigger_array_##name`.

yperess replied:
Thanks, forgot to update since the sensor shell I was using to test just allocated this by size without any values. Fixed.

yperess commented:
> > Right but we're just talking about data representation here. I'm not saying they need to use the subsystem.
> 
> I agree with your points on the data representation, your effort in driving FIFO streaming and vertical decoder are well appreciated. However, our comments on this PR is mainly about the modification of the sensor device driver API itself. With driver API today, an application can work with A/G combo sensor, as well as A/G separated sensor; however, with this PR, this will be broken, and I believe this shall be fixed before we merge this PR.

Lets start off with the current fetch/get model. It's not a good one and there's a reason we want to move away from it. Here's a scenario:

1. Thread 1 calls fetch
    1. Driver caches the data
1. Thread 1 reads ACCEL from cache
2. Context switch
3. Thread 2 calls fetch
    1. Driver caches new data
4.   Context switch
5. Thread 1 reads GYRO (but there's no access to the data from step 1)

The async model that @teburd and I put together puts the data in the ownership of the caller so it looks more like

1. Thread 1 calls fetch
2. Driver allocates a buffer and reads into it, then gives ownership of the buffer to the caller
3. Thread 1 owns the buffer and can release it when it wants
4. Context switch
5. Thread 2 calls fetch
6. Driver repeats step 2 giving the buffer to thread 2
7. Context switch
8. Thread 1 can keep using the same buffer

In my demo that I sent out about the subsytem you can see that the model is that there's a sensor wrapper which is responsible for getting the data from the physical sensor. It's the wrapper that does the demuxing and decoding. Once the demuxing is done based on the original SQE(s), the decoded data is sent as CQE(s) which are handled by the subsystem and sent out via callbacks. Below is a simple example showing a one-shot request. This flow is even more critical in streaming APIs because you don't want the driver to own the data, the options are:
1. Driver owns the data and can't service any more requests until the clients are done
2. Driver supports owning _N_ data buffers simultaneously
3. We implement a shared pointer
4. We have the client own the memory once it's read by the driver <-- this is the approach the async APIs are using.

![Sensors  overview - driver design](https://github.com/zephyrproject-rtos/zephyr/assets/9535482/266fb353-170c-412b-b887-964db706d6ca)

KeHIntel reviewed (changes requested):
To my understanding, the nature of the sensor driver is to offer a common abstraction to the application, and hide the difference from datasheet details. FIFO config is a datasheet detail, it shouldn’t be exposed to the application from sensor driver.

There should always be separated RTIO FIFOs set up for separated type/channel of sensors. The motivation of combo sensor design from sensor vendors are mainly cost saving, Interleaving A/G into one FIFO is a specific HW design choice (and not applicable to all), it does not mean SW should follow.

Furthermore (1), before this PR, I believe a Zephyr application that consumes sensors only needs to care about sensor channel, and does not need to care about any sensor HW characteristics difference. If we approve this PR as is, this assumption will be broken.

Furthermore (2), as the 1st group of users of this API from sensing subsystem, we expect to use the streaming API in the same way for the case of single A/G combo sensor (with one FIFO) vs two standalone A/G sensors (with two FIFOs). I strongly request this PR to consider this requirement.

yperess commented:
> To my understanding, the nature of the sensor driver is to offer a common abstraction to the application, and hide the difference from datasheet details. FIFO config is a datasheet detail, it shouldn’t be exposed to the application from sensor driver.

This API doesn't add any internal data sheet specifications. Users can set the batching using a time duration and the driver will convert that to a FIFO watermark. Is the concern that the triggers being added the `FIFO_WATERMARK` and `FIFO_FULL`? I'm happy to rename them to `BATCH_READY` and `SAMPLE_DROPPED` if you'd like.

> There should always be separated RTIO FIFOs set up for separated type/channel of sensors. The motivation of combo sensor design from sensor vendors are mainly cost saving, Interleaving A/G into one FIFO is a specific HW design choice (and not applicable to all), it does not mean SW should follow.

I get what you're saying, but I don't think it's a reasonable ask. The whole point of the RTIO integration and the async API was to be able to remove all the processing from the drivers. This makes it such that the sensor drivers no longer need to choose between a dedicate processing thread or the system work queue (both weren't great options). Instead, now reading the sensor (both one shot and FIFO) can be done in an interrupt context via the RTIO subsystem that @teburd created). The work of processing the data and splitting it up is left to the client via the `Decoder` API.

> Furthermore (1), before this PR, I believe a Zephyr application that consumes sensors only needs to care about sensor channel, and does not need to care about any sensor HW characteristics difference. If we approve this PR as is, this assumption will be broken.

This PR enables the application to request attributes for batching, if the driver doesn't support it, they'll get an error. This is no different from an application trying to read a channel that's not supported by a sensor currently. We're adding attributes and triggers which may or may not be available on a sensor.

> Furthermore (2), as the 1st group of users of this API from sensing subsystem, we expect to use the streaming API in the same way for the case of single A/G combo sensor (with one FIFO) vs two standalone A/G sensors (with two FIFOs). I strongly request this PR to consider this requirement.

I'm not sure I understand, I'll re-iterate what I understood and please correct me where I'm wrong:

#### Scenario 1
I have 1 sensor with interleaved accel/gyro in the FIFO. Reading this FIFO will give me both sensor types.

I think you're asking the driver to read the FIFO into a buffer, then allocate 2 more buffers and split the data up between the two. If so, this would require a lot of processing that we don't know that the application needs. It would require a processing thread again for every sensor which is what we're trying to move away from. 

#### Scenario 2
I have 2 sensors with separate FIFOs, one sensor provides acceleration and the other gyroscope.

In this case you read acceleration from sensor 0 and gyroscope readings from sensor 1 and there's no processing needed beyond decoding the data.

#### My take on this
Design principals of single responsibility are paramount when designing such a complex and long data pipeline. The sensor driver should not be doing any processing on the data which allows it to run all the reading commands in an interrupt context. It also saves us from doing potentially unnecessary memory copies. I think it's also not reasonable to say that you want to be able to read _JUST_ the accelerometer from scenario 1. Reading the FIFO will clear it, so you'd need to burn a lot of memory to accommodate a use which isn't guaranteed to be needed. Instead, we will provide the caller with the full buffer that was read and a Decoder which allows you to consume the accel and gyro samples independently from the driver. This model works across MCUs even where the decoding isn't done on the same core as the collection. Here' s my view on the responsibilities of the data pipeline stages:

* Driver - collect data and return it to the client
* Decoder - query the data to a format that's compatible with Zephyr's DSP subsystem
* Subsystem - provide support for multiple clients to the same sensor
* Virtual sensors - provide software solutions to sensors that don't physically exist

Adding processing to the driver will overlap the the responsibility of the Decoder.

teburd commented:
I'd add there's maybe a few interesting devices to look at for sort of the extremes of this story. One is basically two devices (accel/gyro) combined into one physical package and register map. The other is a highly integrated sensor hub that interleaves *many* possible events into a single FIFO along with algorithm outputs.

| Device | FIFO(s) | Datasheet |
|--------|--------|--------|
| Bosch BMI085 | Accel FIFO, Gyro FIFO | [Datasheet](https://www.bosch-sensortec.com/media/boschsensortec/downloads/datasheets/bst-bmi085-ds001.pdf) |
| Bosch BHI260AP | Wake-Up FIFO, Non-Wake-Up FIFO, Status and Debug FIFO | [Datasheet](https://www.bosch-sensortec.com/media/boschsensortec/downloads/datasheets/bst-bhi260ap-ds000.pdf) |

It's pretty easy to represent the BMI085 as two distinct devices with two struct device implementations. That's how Zephyr treats it today. Two distinct struct device's.

The second device with a highly integrated sensor hub, including event detection and fusion producing things like orientation vectors would be... incredibly difficult to split into multiple streams of independent data.

yperess commented:
@teburd initial thoughts about both of those. They seem like the extreme cases (which I think was the point). The BMI085 not only has 2 separate FIFOs, but separate interrupts for accel/gyro data ready. The second sensor seems to also provide a lot of extra, I'll call them bells and whistles for now. The driver is suppose to cover the 90/10 rule IMO, the extras can be done via their own `bmi085.h` etc and expose these nuanced functionalities that are specific to the driver. Otherwise, we'd need to add a trigger for every channel and maybe even an index like we did with the decoder. Soon enough the API is just too complicated. I'll dive deeper into these tomorrow.

teburd commented:
Going the opposite route in a thought experiment I had last night, muxing the FIFOs into a single stream seemed to play out reasonably well.

* Streams were already muxed, muxing the FIFOs doesn't change much
* Keeps the single iodev avoiding having to deal with device specifics around how many FIFOs it has
* The decoder still works in the above two bosch devices with multiple FIFOs if the identifier for which FIFO the data was read from is placed in the header just before the read data.

KeHIntel commented:
> I'd add there's maybe a few interesting devices to look at for sort of the extremes of this story. One is basically two devices (accel/gyro) combined into one physical package and register map. The other is a highly integrated sensor hub that interleaves _many_ possible events into several FIFOs along with algorithm outputs.
> 
> Device	FIFO(s)	Datasheet
> Bosch BMI085	Accel FIFO, Gyro FIFO	[Datasheet](https://www.bosch-sensortec.com/media/boschsensortec/downloads/datasheets/bst-bmi085-ds001.pdf)
> Bosch BHI260AP	Wake-Up FIFO, Non-Wake-Up FIFO, Status and Debug FIFO	[Datasheet](https://www.bosch-sensortec.com/media/boschsensortec/downloads/datasheets/bst-bhi260ap-ds000.pdf)

This is exactly the sensor datasheet difference I am talking about, it shouldn't be left to sensor application to handle.

> muxing the FIFOs into a single stream seemed to play out reasonably well.

This is fine to me, as long as the sensor HW detail on FIFO arrangement are hide beneath the streaming API, and no need application level to care, I am OK. Always muxing or always separated are both OK to me.

teburd commented:
> > I'd add there's maybe a few interesting devices to look at for sort of the extremes of this story. One is basically two devices (accel/gyro) combined into one physical package and register map. The other is a highly integrated sensor hub that interleaves _many_ possible events into several FIFOs along with algorithm outputs.
> > Device	FIFO(s)	Datasheet
> > Bosch BMI085	Accel FIFO, Gyro FIFO	[Datasheet](https://www.bosch-sensortec.com/media/boschsensortec/downloads/datasheets/bst-bmi085-ds001.pdf)
> > Bosch BHI260AP	Wake-Up FIFO, Non-Wake-Up FIFO, Status and Debug FIFO	[Datasheet](https://www.bosch-sensortec.com/media/boschsensortec/downloads/datasheets/bst-bhi260ap-ds000.pdf)
> 
> This is exactly the sensor datasheet difference I am talking about, it shouldn't be left to sensor application to handle.
> 
> > muxing the FIFOs into a single stream seemed to play out reasonably well.
> 
> This is fine to me, as long as the sensor HW detail on FIFO arrangement are hide beneath the streaming API, and no need application level to care, I am OK. Always muxing or always separated are both OK to me.

I agree dealing with multiple fifos is exposing the wrong thing.


It seems like the best route then is always muxing.

It sounds like we’ve found agreement then?

KeHIntel commented:
> It seems like the best route then is always muxing.
> 
> It sounds like we’ve found agreement then?

It looks ok to always muxing in the case of combo sensor. However, if there are two standalone sensor, one for A, another for G. In such case, there are two separated sensor device drivers with two separated FIFOs, how can we achieve always muxing?

yperess commented:
@teburd , @tristan-google will be upstreaming the bma422 driver which supports FIFO APIs on our end as the next target.

yperess commented:
> > It seems like the best route then is always muxing.
> > It sounds like we’ve found agreement then?
> 
> It looks ok to always muxing in the case of combo sensor. However, if there are two standalone sensor, one for A, another for G. In such case, there are two separated sensor device drivers with two separated FIFOs, how can we achieve always muxing?

I'm not sure I follow, if you have 2 separate `struct device` objects then you won't mux the 2 streams. We will only mux the two if a single `struct device` has 2+ separate FIFOs.

KeHIntel commented:
> > > It seems like the best route then is always muxing.
> > > It sounds like we’ve found agreement then?
> > 
> > It looks ok to always muxing in the case of combo sensor. However, if there are two standalone sensor, one for A, another for G. In such case, there are two separated sensor device drivers with two separated FIFOs, how can we achieve always muxing?
> 
> I'm not sure I follow, if you have 2 separate `struct device` objects then you won't mux the 2 streams. We will only mux the two if a single `struct device` has 2+ separate FIFOs.

From sensor driver development perspective, this might make sense; but from application perspective, what we want to achieve is to use common code for A/G sensor, no matter it is combo sensor with single FIFO, or multiple FIFOs, or individual sensors with separated FIFOs.

yperess commented:
> > > > It seems like the best route then is always muxing.
> > > > It sounds like we’ve found agreement then?
> > > 
> > > 
> > > It looks ok to always muxing in the case of combo sensor. However, if there are two standalone sensor, one for A, another for G. In such case, there are two separated sensor device drivers with two separated FIFOs, how can we achieve always muxing?
> > 
> > 
> > I'm not sure I follow, if you have 2 separate `struct device` objects then you won't mux the 2 streams. We will only mux the two if a single `struct device` has 2+ separate FIFOs.
> 
> From sensor driver development perspective, this might make sense; but from application perspective, what we want to achieve is to use common code for A/G sensor, no matter it is combo sensor with single FIFO, or multiple FIFOs, or individual sensors with separated FIFOs.

I'm not sure how you'd achieve this if the sensors are in separate device structs. The thinking is that the drivers will provide **ALL** of their data, the decoder will allow you to demux as needed. If you want to have multiple clients, then the suggestion will be to use the subsystem which will own the demuxing and provide you with the ability to subscribe to a single sensor_info node (but that's beyond the scope for the kernel drivers).

KeHIntel commented:
> > > > > It seems like the best route then is always muxing.
> > > > > It sounds like we’ve found agreement then?
> > > > 
> > > > 
> > > > It looks ok to always muxing in the case of combo sensor. However, if there are two standalone sensor, one for A, another for G. In such case, there are two separated sensor device drivers with two separated FIFOs, how can we achieve always muxing?
> > > 
> > > 
> > > I'm not sure I follow, if you have 2 separate `struct device` objects then you won't mux the 2 streams. We will only mux the two if a single `struct device` has 2+ separate FIFOs.
> > 
> > 
> > From sensor driver development perspective, this might make sense; but from application perspective, what we want to achieve is to use common code for A/G sensor, no matter it is combo sensor with single FIFO, or multiple FIFOs, or individual sensors with separated FIFOs.
> 
> I'm not sure how you'd achieve this if the sensors are in separate device structs. The thinking is that the drivers will provide **ALL** of their data, the decoder will allow you to demux as needed. If you want to have multiple clients, then the suggestion will be to use the subsystem which will own the demuxing and provide you with the ability to subscribe to a single sensor_info node (but that's beyond the scope for the kernel drivers).

Talked with Lixu, it looks OK to handle the combo vs individual sensor case in phy_3d_sensor_wrapper, if combo sensor always has muxing FIFO, and individual sensor always has separate FIFO.

lixuzha commented on include/zephyr/drivers/sensor.h:
```diff
@@ -567,12 +611,43 @@ struct sensor_read_config {
+ * @endcode
+ */
+#define SENSOR_DT_STREAM_IODEV(name, dt_node, ...)                                                 \
+	static struct sensor_stream_trigger __trigger_array_##name[] = {__VA_ARGS__};              \
```
Hi Yuval,

I met some problem when I integrated this PR. 
It would be better using _CONCAT instead of `##` when this macro is used by another macro.

yperess replied:
Done

lixuzha replied:
Thanks. Please also use `_CONCAT` in `SENSOR_DT_READ_IODEV`.

yperess commented:
@MaureenHelm @KeHIntel @lixuzha can you re-review?

lixuzha commented:
> @MaureenHelm @KeHIntel @lixuzha can you re-review?

Please also use `_CONCAT` in `SENSOR_DT_READ_IODEV`. Thanks.

yperess commented:
CI is blocked on https://github.com/zephyrproject-rtos/zephyr/pull/64881

nashif commented:
> CI is blocked on #64881

can you rebase please?

nashif commented:
> can you rebase please?

oh, it is passing already, why is this considered blocked

kartben commented:
> > can you rebase please?
> 
> oh, it is passing already, why is this considered blocked

@MaureenHelm has requested changes - but it looks like they have been addressed.
